### PR TITLE
feat: configure sketchbook location without restart

### LINKED ISSUE #569: If Sketchbook is empty, no Sketchbook menu item is present under File
### Describe the problem

This happens in two occurrences:

1. If you have a brand new installation there is no Sketchbook item under File
2. If you point to an empty folder as your Sketchbook location and restart the IDE without saving a single Sketch, when you relaunch the IDE you can observe the same behaviour

### To reproduce

1. Select **File > Preferences** from the Arduino IDE menus.
1. Change the "**Sketchbook location** preference to some folder that does not contain sketches.
1. Click the <kbd>OK</kbd> button.
1. Select **File > New** from the Arduino IDE menus.
   **ⓘ** This is necessary to workaround a bug where the IDE doesn't update the menus after the preference change: https://github.com/arduino/arduino-ide/issues/796
1. Select **File** from the Arduino IDE menus.

🐛 There is no **File > Sketchbook** menu:

image

### Expected behavior

The "**Sketchbook**" item should always be there even if empty, as it works in Arduino IDE 1.8.x:

image

### Arduino IDE version

2.0.0 Beta 12

### Operating system

macOS (probably any OS)

### Operating system version

Big Sur

### Additional context

Related: https://github.com/arduino/arduino-ide/issues/655

### Issue checklist

- [X] I searched for previous reports in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the problem still occurs when using the latest [nightly build](https://github.com/arduino/arduino-ide#nightly-builds)
- [X] My report contains all necessary details

### LINKED ISSUE #655: Show "Tools > Port" menu even when empty
### Describe the request

Show A **Port** menu item under the Arduino IDE **Tools** menu even when there are no ports to populate that **Tools > Port** menu.

🙂 The lack of ports will be communicated to the user in an intuitive manner.

### Describe the current behavior

The Arduino IDE's **Tools > Port** menu is populated by the discovered ports.

An unexpected lack of a port is a common problem for Arduino users. This might happen when the user has a bad USB cable, didn't install the drivers for their board, burned up the board, etc.

Currently, when no ports were discovered the **Tools > Port** menu item is missing entirely. I feel this is not a great user experience. An empty menu better communicates the lack of a port to the user than the menu itself being absent. The missing menu gives the user the impression there is a fault in the Arduino IDE application rather than a lack of an external resource listed by the menu.

image

Compare to the behavior of Arduino IDE 1.x under these conditions:

image

### Arduino IDE version

2.0.0-rc1-snapshot.f0d9894

### Operating system

Windows

### Operating system version

10

### Additional context

There are some other menus that are dependent on external resources:

- **File > Sketchbook** - Equivalent request tracked at https://github.com/arduino/arduino-ide/issues/569
- **File > Examples** - this will never be empty under normal conditions due to the built-in examples included with the Arduino IDE installation
- **Tools > Board** - this is unlikely to be empty due to the IDE installing Arduino AVR Boards on first run (https://github.com/arduino/arduino-ide/pull/497)

---

Originally reported at https://forum.arduino.cc/t/compilation-fails-due-to-no-upload-port-specified/926836/5

---

Additional reports:

- https://forum.arduino.cc/t/why-is-there-no-port-menu-item-under-tools/942983

### LINKED ISSUE #796: IDE UI does not reflect changed "Sketchbook location" preference
### Describe the problem

User sketches and global library installations are stored in the `libraries` subfolder of the Arduino sketchbook folder. Several UI elements of the Arduino IDE reflect the contents of this folder and the build system also does library discovery from it.

The user may change the location of this sketchbook folder using the **Sketchbook location** preference of the Arduino IDE.

:bug: After the user changes the sketchbook location preference, the IDE's UI elements do not reflect the change to sketchbook sketches and available libraries.

### To reproduce

1. Select **File > New** from the Arduino IDE menus.
1. Select **Tools > Board > Arduino AVR Boards > Arduino Uno** from the Arduino IDE menus.
   **ⓘ** This is done to avoid variation in results  that could be caused by [platform bundled libraries](https://arduino.github.io/arduino-cli/dev/platform-specification/#platform-bundled-libraries) of an arbitrary board selection.
1. Open the "**Sketchbook**" view.
1. Take note of the sketches listed there.
1. Select **File > Sketchbook** from the Arduino IDE menus.
1. Take note of the sketches listed there.
1. Select **File > Examples** from the Arduino IDE menus.
1. Take note of the libraries listed there.
   :warning: There is a separate issue https://github.com/arduino/arduino-ide/issues/262 which intermittently causes this menu to be incomplete. If you don't see the libraries you have installed listed in the menu (e.g., "**Ethernet**", "**Firmata**", "**Keyboard**"), then try selecting **File > New** to give the IDE another chance to populate the menu correctly.
1. Select **Sketch > Include Library** from the Arduino IDE menus.
1. Take note of the libraries listed there.
1. Select **File > Preferences...** from the Arduino IDE menus.
1. Change the **Sketchbook location** preference to some other folder on your computer which already exists, and which contains different sketches and libraries (none at all is fine) from the current sketchbook folder.
   **ⓘ** The bug occurs regardless of whether you set the new sketchbook path directly via the **Sketchbook location** field or using the dialog accessed by clicking the <kbd>**Browse**</kbd> button.
1. Click the <kbd>**OK**</kbd> button.
1. Open the "**Sketchbook**" view.
   :bug: The view shows the contents of the previous sketchbook location.
1. Select **File > Sketchbook** from the Arduino IDE menus.
   :bug: The menu shows the sketches located in the previous sketchbook location.
   :bug: The menu does not show the sketches located in the current sketchbook location.
1. Select **File > Examples** from the Arduino IDE menus.
   :bug: The menu shows the examples of the libraries from the previous sketchbook location.
   :bug: The menu does not show the examples of the libraries from the current sketchbook location.
1. Select **Sketch > Include Library** from the Arduino IDE menus.
   :bug: The menu shows the libraries from the previous sketchbook location.
   :bug: The menu does not show the libraries from the current sketchbook location.
1. Select one of the unexpected libraries from the menu.
   `#include` directive(s) for the library are added to the sketch.
1. Select **Sketch > Verify/Compile** from the Arduino IDE menus.
   :bug: The compilation fails with a "`No such file or directory`" error, since the library is not in one of the libraries paths.
1. Search for the library in Library Manager.
   ~~:bug: Library Manager shows the library as installed.~~
   **UPDATE**: This specific part was resolved by https://github.com/arduino/arduino-ide/pull/1132. However, that resolution is unexpected since the index update (which triggers the reinitialization of Arduino CLI that updates the library installation status) should only be triggered on "**Additional Boards Manager URLs**" preference change. So the fact that the index update is currently (fc9107c at the time this is being written) being triggered on "**Sketchbook location**" preference change is a bug.
1. Install a library via Library Manager.
   ~~:bug: The library is installed to the `libraries` subfolder of the old sketchbook location.~~
   **UPDATE**: This specific part has been resolved since the time of the report (though perhaps as a side effect of a bug, as described in the note regarding the Library Manager installation status).
1. Install a library via **Sketch > Include Library > Add .ZIP Library...**
   ~~:bug: The library is installed to the `libraries` subfolder of the old sketchbook location.~~
   **UPDATE**: This specific part has been resolved since the time of the report (though perhaps as a side effect of a bug, as described in the note regarding the Library Manager installation status).

### Expected behavior

All IDE data on sketchbook contents (sketches, libraries, manually installed boards platforms) is updated after changing the sketchbook location preference.

#### Original report

2.0.0-rc3-snapshot.759b410

#### Last verified with

2.0.0-snapshot-42f6f43

### Operating system

Windows

### Operating system version

10

### Additional context

In addition to the use case of the **Sketchbook location** preference for occasional long term change of the sketchbook location according to user whims, some advanced users dedicate a separate sketchbook to each project in order to provide it with a dedicated dependencies environment, which may include manually installed boards platforms in addition to libraries. These users frequently switch from one sketchbook to another and so are especially dependent on the IDE immediately recognizing that change.

---

~~The issue also applies to boards platforms manually installed to the `hardware` subfolder of the sketchbook.~~

**UPDATE**: This specific part has been resolved since the time of the report (though perhaps as a side effect of a bug, as described in the note regarding the Library Manager installation status).

---

Possibly related to https://github.com/arduino/arduino-cli/issues/1802

### Issue checklist

- [X] I searched for previous requests in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the feature was still missing when using the latest [nightly build](https://github.com/arduino/arduino-ide#nightly-builds)
- [X] My request contains all necessary details

### LINKED ISSUE #1764: IDE hangs on startup if drive containing sketchbook location not accessible
### Describe the problem

The Arduino sketchbook is a convenient place to store sketches and the location where Arduino IDE installs libraries.

The user can set a custom location for the sketchbook via the "**Sketchbook location**" preference.

🐛 If the drive containing the "**Sketchbook location**" path is not available, the Arduino IDE startup hangs. The cause is not communicated to the user in any way.

#### Setup

1. Connect a removable drive to your computer.
1. Select **File > Preferences...** from the Arduino IDE menus.
1. Set a path under the removable drive in the "**Sketchbook location**" field.
1. Click the <kbd>**OK**</kbd> button.
1. Select **File > Quit** from the Arduino IDE menus.

-OR-

The fault can also be reproduced even if you don't have a removable drive on hand:

1. Select **File > Quit** from the Arduino IDE menus.
1. Open the file at the following path in a text editor:
   - **Windows**:
     ```text
     C:\Users\<username>\AppData\Roaming\arduino-ide\arduino-cli.yaml
     ```
   - **Linux**:
     ```text
     ~/.config/arduino-ide/arduino-cli.yaml
     ```
   - **macOS**:
     ```text
     ~/Library/Application Support/arduino-ide/arduino-cli.yaml
     ```
1. Change the path in the `directories.user` key to one under a drive that doesn't exist on your computer.
   For example:
   ```yaml
   user: L:\Arduino
   ```
1. Save the file.

#### Steps

1. Start Arduino IDE.

🐛 Startup hangs at the loading animation:

image

### Expected behavior

Arduino IDE clearly communicates to the user that the sketchbook location is inaccessible. If it is not possible for the IDE to be functional under these conditions, this should also be clearly communicated to the user and they should still be able to change the "**Sketchbook location**" preference to an accessible path to restore the functionality.

### Arduino IDE version

f8c01e3

### Operating system

Windows

### Operating system version

10

### Additional context

I bisected the introduction of the bug to https://github.com/arduino/arduino-ide/commit/6983c5bf7f40de76542de0c4a0810bcd3f204612 (does not occur with the build for the previous commit b3ab5cb)

---

I see this in the logs on startup:

```text
2022-12-12T06:43:22.393Z root ERROR Uncaught Exception:
2022-12-12T06:43:22.393Z root ERROR Error: ENOENT: no such file or directory, mkdir '\\?'
```

---

The fault does not occur if the drive is accessible but the sketchbook folder is missing. In this case, the missing folder is simply created on startup (https://github.com/arduino/arduino-ide/commit/6983c5bf7f40de76542de0c4a0810bcd3f204612).

---

Originally reported at https://forum.arduino.cc/t/arduino-ide-2-wont-start/1063769

### Issue checklist

- [X] I searched for previous reports in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the problem still occurs when using the latest [nightly build](https://www.arduino.cc/en/software#nightly-builds)
- [X] My report contains all necessary details

### PR DESCRIPTION
**Pull request by kittaakos**
### Motivation

 - Handle invalid `directories.user` path in `arduino-cli.yaml` at app startup.
 - Change the sketchbook location from IDE2 without a restart.
 - Show the `File` > `Sketchbook` menu item even if no sketches are in the folder.
 - Show the `Tools` > `Port` menu item even if there are no available ports.

Preview:

https://user-images.githubusercontent.com/1405703/207853423-3c1e3502-9904-49d2-b2af-8ef01367af19.mp4

### Change description

 - IDE2 can start when `directories.user` is invalid or does not exist.
 - If `directories.user` is not a valid directory, IDE2 toasts an error message at startup.
 - `Save As...`, `Open`, `Archive Sketch`, and `Add .ZIP Library...` are fully functional when there is no valid `directories.user` value is set.
 - If `directories.user` does not exist, IDE2 tries to create the folder.
 - It's possible to change the `directories.user` location from the IDE2's _Settings_ dialog.
 - When the user changes the `directories.user`, `File` > `Sketchbook`, `File` > `Examples`, and `Sketch` > Include Library` submenus update.
 - When the `directories.user` is invalid or empty, `File` > `Sketchbook` is visible but disabled.
 - `Tools` > `Port` is visible but disabled when no available ports are detected.

### Other information

Closes #1764
Closes #796
Closes #569
Closes #655

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: arduino-ide-extension/src/browser/contributions/archive-sketch.ts [hunk 1]
```diff
@@ -29,10 +29,12 @@ export class ArchiveSketch extends SketchContribution {
   }
 
   private async archiveSketch(): Promise<void> {
-    const [sketch, config] = await Promise.all([
-      this.sketchServiceClient.currentSketch(),
-      this.configService.getConfiguration(),
-    ]);
+    const config = this.configService.tryGetConfig();
+    if (!config) {
+      return;
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/contributions/archive-sketch.ts:34]: TODO: fallback to any path.

File: arduino-ide-extension/src/browser/contributions/open-sketch.ts [hunk 2]
```diff
@@ -82,7 +82,10 @@ export class OpenSketch extends SketchContribution {
   }
 
   private async selectSketch(): Promise<Sketch | undefined> {
-    const config = await this.configService.getConfiguration();
+    const config = this.configService.tryGetConfig();
+    if (!config) {
+      return undefined;
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/contributions/open-sketch.ts:87]: Incorrect. Opening a sketch should work even if the sketchbook path is invalid/not set.

File: arduino-ide-extension/src/browser/contributions/save-as-sketch.ts [hunk 3]
```diff
@@ -58,10 +58,11 @@ export class SaveAsSketch extends SketchContribution {
       markAsRecentlyOpened,
     }: SaveAsSketch.Options = SaveAsSketch.Options.DEFAULT
   ): Promise<boolean> {
-    const [sketch, configuration] = await Promise.all([
-      this.sketchServiceClient.currentSketch(),
-      this.configService.getConfiguration(),
-    ]);
+    const configuration = this.configService.tryGetConfig();
+    if (!configuration) {
+      return false;
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/contributions/save-as-sketch.ts:63]: Fix this

File: arduino-ide-extension/src/browser/contributions/sketchbook.ts [hunk 4]
```diff
@@ -1,44 +1,55 @@
...
-    this.registerRecursively(
-      [...container.children, ...container.sketches],
-      ArduinoMenus.FILE__SKETCHBOOK_SUBMENU,
-      this.toDispose
+    this.toDispose.push(
+      Disposable.create(() =>
+        this.menuRegistry.unregisterMenuNode(
+          ArduinoMenus.FILE__SKETCHBOOK_SUBMENU[
+            ArduinoMenus.FILE__SKETCHBOOK_SUBMENU.length - 1
+          ]
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/contributions/sketchbook.ts:33]: TODO: code-comment that this assumes how Theia calculates the submenu ID from the menu path.

File: arduino-ide-extension/src/browser/dialogs/settings/settings.ts [hunk 5]
```diff
@@ -169,9 +169,17 @@ export class SettingsService {
...
-    const { additionalUrls, sketchDirUri, network } = cliConfig;
+    const { config } = cliConfig;
+    let additionalUrls: string[] = [];
+    let sketchDirUri = '';
+    let network = Network.Default();
+    if (config) {
+      additionalUrls = config.additionalUrls;
+      sketchDirUri = config.sketchDirUri;
+      network = config.network;
+    }
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/dialogs/settings/settings.ts:182]: cleanup

File: arduino-ide-extension/src/browser/local-cache/local-cache-fs-provider.ts [hunk 6]
```diff
@@ -88,7 +88,10 @@ export class LocalCacheFsProvider
   }
 
   protected async init(fileService: FileService): Promise<void> {
-    const config = await this.configService.getConfiguration();
+    const { config, messages } = await this.configService.config();
+    if (!config) {
+      throw new Error(messages ? messages[0] : 'cli config was undefined');
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/local-cache/local-cache-fs-provider.ts:93]: check the typings here. increase TS level? if config is not defined, messages must be defined

File: arduino-ide-extension/src/common/protocol/config-service.ts [hunk 7]
```diff
@@ -6,10 +6,13 @@ export interface ConfigService {
...
-  getCliConfigFileUri(): Promise<string>;
-  getConfiguration(): Promise<Config>;
-  setConfiguration(config: Config): Promise<void>;
+  configFileUri(): Promise<string>;
+  config(): Promise<ConfigState>;
+  updateConfig(config: Config): Promise<void>;
 }
+export type ConfigState =
+  | { config: undefined; messages: string[] }
+  | { config: Config; messages?: string[] };
```
Comment (kittaakos) [inline arduino-ide-extension/src/common/protocol/config-service.ts:15]: counterpart of the TS version related comment

File: arduino-ide-extension/src/electron-browser/theia/core/electron-main-menu-factory.ts [hunk 8]
```diff
@@ -178,9 +178,13 @@ export class ElectronMainMenuFactory extends TheiaElectronMainMenuFactory {
   ): Electron.MenuItemConstructorOptions[] {
     const showDisabled = options?.showDisabled !== false;
 
+    if (CompoundMenuNode.is(menu) && menu.id.includes('2_ports')) {
+      debugger;
+    }
```
Comment (kittaakos) [inline arduino-ide-extension/src/electron-browser/theia/core/electron-main-menu-factory.ts:183]: trash

File: arduino-ide-extension/src/electron-browser/theia/core/electron-main-menu-factory.ts [hunk 9]
```diff
@@ -178,9 +178,13 @@ export class ElectronMainMenuFactory extends TheiaElectronMainMenuFactory {
...
+
     if (
       CompoundMenuNode.is(menu) &&
-      menu.children.length &&
+      (menu.children.length || alwaysVisibleSubmenu(menu)) && // hack for #569 and #655
```
Comment (kittaakos) [inline arduino-ide-extension/src/electron-browser/theia/core/electron-main-menu-factory.ts:187]: This should go upstream, without the hack
Also propose a change with the enablements

File: arduino-ide-extension/src/node/clang-formatter.ts [hunk 10]
```diff
@@ -80,8 +80,11 @@ export class ClangFormatter implements Formatter {
   }
 
   private async dataDirPath(): Promise<string> {
-    const { dataDirUri } = await this.configService.getConfiguration();
-    return FileUri.fsPath(dataDirUri);
+    const { config, messages } = await this.configService.config();
+    if (!config) {
+      throw new Error(messages ? messages[0] : 'cli config was undefined');
```
Comment (kittaakos) [inline arduino-ide-extension/src/node/clang-formatter.ts:85]: caller should handle undefined, formatter must still work without a valid data dir

File: arduino-ide-extension/src/node/sketches-service-impl.ts [hunk 11]
```diff
@@ -184,8 +184,11 @@ export class SketchesServiceImpl
   }
 
   private async sketchbookUri(): Promise<string> {
-    const { sketchDirUri } = await this.configService.getConfiguration();
-    return sketchDirUri;
+    const { config, messages } = await this.configService.config();
+    if (!config) {
+      throw new Error(messages ? messages[0] : 'cli config was undefined');
```
Comment (kittaakos) [inline arduino-ide-extension/src/node/sketches-service-impl.ts:189]: must be handled in getSketches

File: arduino-ide-extension/src/browser/sketches-service-client-impl.ts [hunk 12]
```diff
@@ -0,0 +1,296 @@
+import { inject, injectable } from '@theia/core/shared/inversify';
```
[the remark this answers is not in the document]
Comment (kittaakos) [reply arduino-ide-extension/src/browser/sketches-service-client-impl.ts]: https://github.com/arduino/arduino-ide/pull/1774/commits/35d9fe13ad4132adfc8d3be106e2cfdbac4c7bbb

File: arduino-ide-extension/src/browser/contributions/sketchbook.ts (hunk 4 above)
Comment (kittaakos) [reply arduino-ide-extension/src/browser/contributions/sketchbook.ts:33]: https://github.com/arduino/arduino-ide/pull/1774/commits/40bdae69e5f9c0f814a41565aac9fc804b91dafa

File: arduino-ide-extension/src/electron-browser/theia/core/electron-main-menu-factory.ts (hunk 8 above)
Comment (kittaakos) [reply arduino-ide-extension/src/electron-browser/theia/core/electron-main-menu-factory.ts]: https://github.com/arduino/arduino-ide/pull/1774/commits/157305cba1ed1fbf45615ea89b0d7a5f52275bcc

File: arduino-ide-extension/src/browser/theia/monaco/monaco-text-model-service.ts [hunk 13]
```diff
@@ -6,14 +6,16 @@ import { EditorPreferences } from '@theia/editor/lib/browser/editor-preferences'
...
 
 @injectable()
 export class MonacoTextModelService extends TheiaMonacoTextModelService {
   @inject(SketchesServiceClientImpl)
   protected readonly sketchesServiceClient: SketchesServiceClientImpl;
 
-  protected override async createModel(resource: Resource): Promise<MonacoEditorModel> {
+  protected override async createModel(
+    resource: Resource
+  ): Promise<MonacoEditorModel> {
```
[the remark this answers is not in the document]
Comment (kittaakos) [reply arduino-ide-extension/src/browser/theia/monaco/monaco-text-model-service.ts]: https://github.com/arduino/arduino-ide/pull/1774/commits/20c775c4dc37e21326380ce078d261785bdfe0af

File: arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts [hunk 14]
```diff
@@ -1,8 +1,12 @@
-import { inject, injectable, postConstruct } from '@theia/core/shared/inversify';
+import {
+  inject,
+  injectable,
+  postConstruct,
+} from '@theia/core/shared/inversify';
```
[the remark this answers is not in the document]
Comment (kittaakos) [reply arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts]: https://github.com/arduino/arduino-ide/pull/1774/commits/22c3a4da226f4ca16d341b4d8ddd32e51c8edcb4

File: arduino-ide-extension/src/browser/contributions/open-sketch.ts (hunk 2 above)
Comment (kittaakos) [reply arduino-ide-extension/src/browser/contributions/open-sketch.ts]: https://github.com/arduino/arduino-ide/pull/1774/commits/707f59dbc9594738655f4b2d2d4c03e317c3e384

File: arduino-ide-extension/src/browser/contributions/archive-sketch.ts (hunk 1 above)
Comment (kittaakos) [reply arduino-ide-extension/src/browser/contributions/archive-sketch.ts:33]: https://github.com/arduino/arduino-ide/pull/1774/commits/ae7429a257673879ee1c8b62b9cc79ffce42b6e6

Comment (kittaakos) [reply arduino-ide-extension/src/browser/contributions/archive-sketch.ts]: https://github.com/arduino/arduino-ide/pull/1774/commits/784c911d54afc22a023cf617ed0c8700dfe6466c

File: arduino-ide-extension/src/browser/contributions/save-as-sketch.ts (hunk 3 above)
Comment (kittaakos) [reply arduino-ide-extension/src/browser/contributions/save-as-sketch.ts]: https://github.com/arduino/arduino-ide/pull/1774/commits/16abc5555ed30a9ed6af7ba04a30efc30dd18f13

File: arduino-ide-extension/src/node/clang-formatter.ts (hunk 10 above)
Comment (kittaakos) [reply arduino-ide-extension/src/node/clang-formatter.ts]: https://github.com/arduino/arduino-ide/pull/1774/commits/30c4af958cab9ffb3672a6c26bd13d6771a95291

File: arduino-ide-extension/src/browser/local-cache/local-cache-fs-provider.ts (hunk 6 above)
Comment (kittaakos) [reply arduino-ide-extension/src/browser/local-cache/local-cache-fs-provider.ts:93]: > local cache provider must re-init on data dir config change

WONTDO: there is no way in IDE2 to change the data dir. Hence, handling this change is outside of the scope of the PR.

File: arduino-ide-extension/src/browser/dialogs/settings/settings.ts (hunk 5 above)
Comment (kittaakos) [reply arduino-ide-extension/src/browser/dialogs/settings/settings.ts]: https://github.com/arduino/arduino-ide/pull/1774/commits/f660ae5869f2b1a123b442c5d94b94f79c1d7b7c

File: arduino-ide-extension/src/node/sketches-service-impl.ts (hunk 11 above)
Comment (kittaakos) [reply arduino-ide-extension/src/node/sketches-service-impl.ts]: https://github.com/arduino/arduino-ide/pull/1774/commits/23c2791c814d0be5d04b3f1afc1b661a8605f998

File: arduino-ide-extension/src/common/protocol/config-service.ts (hunk 7 above)
Comment (kittaakos) [reply arduino-ide-extension/src/common/protocol/config-service.ts:15]: It has been resolved on call sites.

File: arduino-ide-extension/src/electron-browser/theia/core/electron-main-menu-factory.ts (hunk 9 above)
Comment (kittaakos) [reply arduino-ide-extension/src/electron-browser/theia/core/electron-main-menu-factory.ts]: https://github.com/arduino/arduino-ide/pull/1774/commits/b0b8f7781eebdf65d035277722f8579ceaef2d0d

**Review by per1234 — APPROVED** [review]
I verified this fixes:

- https://github.com/arduino/arduino-ide/issues/1764
- https://github.com/arduino/arduino-ide/issues/796
- https://github.com/arduino/arduino-ide/issues/569
- https://github.com/arduino/arduino-ide/issues/655

Thanks Akos!

**Pull request merged**